### PR TITLE
[Windows] Re-implement Windows support using C++

### PR DESCRIPTION
**Pull request by tritao**
This PR removes the existing C# implementation of the module and updates it with an almost literal C++ port with some updates to use JSValue functionality instead of JSON types internally.

This work is being done on behalf of Microsoft in an effort to update modules.

### CODE REVIEW

File: windows/RNTrackPlayer/Logic/Utils.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+winrt::Windows::Foundation::Uri Utils::GetUri(const React::JSValueObject& obj,
+    std::string key, winrt::Windows::Foundation::Uri def)
+{
+    const JSValue& val = obj[key];
+    if (val == JSValue::Null) return def;
+
+    if (val.Type() == JSValueType::Object)
+        return Uri(winrt::to_hstring(val.AsObject()["uri"].AsString()));
```
Comment (asklar) [inline windows/RNTrackPlayer/Logic/Utils.cpp:30]: what if uri is not defined in the object? (probably want to fail)

**Review by asklar — CHANGES_REQUESTED** [review]
doing a first pass : ) main feedback is: 
- discourage the use of raw pointers (use shared_ptr and the likes for relationships where two objects need to know about each other, or identify whether they can be merged into the same object)
- clean up the private/public boundary for these structs
- take input strings (or any complex type) by ref
- use constexpr for constant values known at compile time
- error checking missing in some cases, consider providing meaningful error messages

File: windows/RNTrackPlayer/Logic/Metadata.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+#include "Logic/Track.h"
+#include "Logic/Utils.h"
+
+using namespace winrt::RNTrackPlayer;
+using namespace winrt::Windows::Media;
+using namespace winrt::Windows::Storage::Streams;
+
+Metadata::Metadata(MediaManager* manager)
+    : controls(nullptr)
+    , jumpInterval(15)
```
Comment (rectified95) [inline windows/RNTrackPlayer/Logic/Metadata.cpp:14]: (nit) Slightly odd formatting - move the commas to end of line?

File: windows/RNTrackPlayer/Logic/Metadata.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    if (data.find("jumpInterval") != data.end())
+    {
+        jumpInterval = data["jumpInterval"].AsDouble();
+    }
+
+    if (data.find("capabilities") != data.end())
+    {
+        auto& capabilities = data["capabilities"].AsArray();
+
+        play = Utils::ContainsInt(capabilities, (int)Capability::Play);
```
Comment (rectified95) [inline windows/RNTrackPlayer/Logic/Metadata.cpp:86]: (nit) prefer static_cast in modern c++

File: windows/RNTrackPlayer/Logic/Metadata.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        pause = Utils::ContainsInt(capabilities, (int)Capability::Pause);
+        stop = Utils::ContainsInt(capabilities, (int)Capability::Stop);
+        previous = Utils::ContainsInt(capabilities, (int)Capability::Previous);
+        next = Utils::ContainsInt(capabilities, (int)Capability::Next);
+        jumpForward = Utils::ContainsInt(capabilities, (int)Capability::JumpForward);
+        jumpBackward = Utils::ContainsInt(capabilities, (int)Capability::JumpBackward);
+        seek = Utils::ContainsInt(capabilities, (int)Capability::Seek);
+
+        if (controls != nullptr) UpdateCapabilities();
```
Comment (rectified95) [inline windows/RNTrackPlayer/Logic/Metadata.cpp:95]: (nit) use curly brackets for consistent styling of if-blocks

File: windows/RNTrackPlayer/Logic/Metadata.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+void Metadata::UpdateMetadata(Track& track)
+{
+    auto display = controls.DisplayUpdater();
+    display.AppMediaId(winrt::to_hstring(track.Id));
+
+    try
+    {
+
```
Comment (rectified95) [inline windows/RNTrackPlayer/Logic/Metadata.cpp:106]: remove blank line

File: windows/RNTrackPlayer/Logic/Metadata.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        Seek = 6,
+        JumpForward = 7,
+        JumpBackward = 8
+    };
+
+    struct MediaManager;
+    struct Track;
+
+    struct Metadata {
+        MediaManager* manager;
```
[the remark this answers is not in the document]
Comment (rectified95) [reply windows/RNTrackPlayer/Logic/Metadata.h:23]: +1 on this, use a pair of std:shared_ptr and std::weak_ptr

File: windows/RNTrackPlayer/RNTrackPlayer.cpp [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+void TrackPlayerModule::Remove(JSValueArray arr, ReactPromise<JSValue> promise) noexcept
+{
+    auto player = manager ? manager->GetPlayer() : nullptr;
+    if (Utils::CheckPlayback(player, promise))
+        return;
+
+    std::vector<std::string> tracks;
+    tracks.reserve(arr.size());
+
+    for(const JSValue& id : arr)
```
Comment (rectified95) [inline windows/RNTrackPlayer/RNTrackPlayer.cpp:166]: missing space for" "(

**Comment by tritao** [comment]
Updated the PR with the review feedback.

I opted to use a `std::unique_ptr` since there is no need for shared ownership `std::shared_ptr`, and there is also no need for the lifecycle tracking feature of the `weak_ptr`. Hope that's OK, if not, I can change it.

Let me know if there is anything else.

**Comment by tritao** [comment]
Upon further testing, found another couple issues. 

Pushed fixes for incorrect duration calcs as well as the loading callback not working.

**Comment by asklar** [comment]
@tritao please let us know when it's ready to review again : )

**Comment by tritao** [comment]
It should be ready to go now.

**Comment by dcvz** [comment]
thank you for this contribution @tritao! I will try this out soon and see how to best get it into our ongoing efforts of v2.

**Comment by dcvz** [comment]
@tritao it looks pretty solid to me although my c++ is not super amazing.
Eventually it would be great to add windows to the example app we have in the repository that can be validated by the CI.

Currently we're working hard to get v2 shipped and released.. in there we introduced one big change vs the current version which is we changed the queue modification methods: `add`, `remove`, `skipTo`, etc.. to be index based instead of ID based.

Would you be interested in lining this up with our current `main` branch (where we're working on v2) and change the methods to be index based?

**Comment by tritao** [comment]
@dcvz Sorry for delay in getting back to you, had to discuss with manager.

I see you already opened another PR with some WIP, so I'll close this one.

**Comment by dcvz** [comment]
Followed up in the new PR @tritao could still benefit from some C++ experience to get this effort in

**Comment by stmoy** [comment]
Linking up the new PR: https://github.com/DoubleSymmetry/react-native-track-player/pull/1180

**Pull request closed without merging**